File: client/cli/src/chat/chatManager.ts
/**
 * Chat message management — encrypt, send, receive, decrypt, store.
 *
 * Messages are encrypted using the session key established during the
 * handshake and signed with the identity signing key.
 */

import { v4 as uuidv4 } from 'uuid';
import { encryptMessage, decryptMessage } from '@aicq/crypto';
import { ClientStore } from '../store.js';
import { P2PClient } from '../p2p/p2pClient.js';
import { WSClient } from '../services/wsClient.js';
import { IdentityManager } from '../services/identityManager.js';
import type { ChatMessage, FileMetadata } from '../types.js';

export class ChatManager {
  private store: ClientStore;
  private p2p: P2PClient;
  private ws: WSClient;
  private identity: IdentityManager;

  /** Debounced save timer to avoid blocking the event loop on hot paths. */
  private _saveTimer: ReturnType<typeof setTimeout> | null = null;

  /** Emitter-like callbacks for external consumers. */
  private messageCallbacks: ((msg: ChatMessage) => void)[] = [];

  constructor(
    store: ClientStore,
    p2p: P2PClient,
    ws: WSClient,
    identity: IdentityManager,
  ) {
    this.store = store;
    this.p2p = p2p;
    this.ws = ws;
    this.identity = identity;

    this._setupListeners();
  }

  /* ──────────────── Callbacks ──────────────── */

  onMessage(callback: (msg: ChatMessage) => void): void {
    this.messageCallbacks.push(callback);
  }

  private notifyMessage(msg: ChatMessage): void {
    for (const cb of this.messageCallbacks) {
      try {
        cb(msg);
      } catch {
        // Ignore callback errors
      }
    }
  }

  /* ──────────────── WebSocket listeners ──────────────── */

  private _setupListeners(): void {
    // Listen for incoming messages via WebSocket fallback
    this.ws.on('message', (msg: any) => {
      if (msg.data?.type === 'encrypted_message') {
        const message = this.receiveMessage(
          Buffer.from(msg.data.payload, 'base64'),
          msg.from ?? msg.data.fromId,
        );
        if (message) {
          this.notifyMessage(message);
        }
      }
    });

    // Listen for P2P data messages
    this.p2p.onAnyData((peerId: string, data: Buffer) => {
      const message = this.receiveMessage(data, peerId);
      if (message) {
        this.notifyMessage(message);
      }
    });
  }

  /* ──────────────── Send ──────────────── */

  /**
   * Send a text message to a friend.
   *
   * 1. Check friend exists
   * 2. Get session key
   * 3. Encrypt with session key + sign with identity key
   * 4. Send via P2P (preferred) or WS fallback
   * 5. Store in history
   */
  async sendMessage(friendId: string, content: string): Promise<ChatMessage> {
    const sessionKey = this.store.getSessionKey(friendId);
    if (!sessionKey) {
      throw new Error(`No session key for friend ${friendId}`);
    }

    // Create message record
    const message: ChatMessage = {
      id: uuidv4(),
      fromId: this.identity.getUserId(),
      toId: friendId,
      type: 'text',
      content,
      timestamp: Date.now(),
      status: 'sent',
    };

    // Encrypt and sign
    const wireData = encryptMessage(
      content,
      sessionKey,
      this.identity.getSigningSecretKey(),
      this.identity.getPublicKey(),
    );

    // Send via P2P if connected, otherwise fall back to WebSocket
    let delivered = false;
    if (this.p2p.isConnected(friendId)) {
      delivered = this.p2p.send(friendId, Buffer.from(wireData));
    }

    if (!delivered) {
      // WebSocket fallback
      this.ws.send('message', {
        to: friendId,
        data: {
          type: 'encrypted_message',
          fromId: this.identity.getUserId(),
          payload: Buffer.from(wireData).toString('base64'),
        },
      });
    }

    message.status = delivered ? 'delivered' : 'sent';
    this.store.addMessage(friendId, message);
    this._debouncedSave();

    return message;
  }

  /**
   * Send file metadata (file-info message) to a friend.
   * This tells the peer about an upcoming file transfer.
   */
  async sendFileInfo(
    friendId: string,
    fileInfo: FileMetadata,
  ): Promise<ChatMessage> {
    const content = JSON.stringify(fileInfo);

    const message: ChatMessage = {
      id: uuidv4(),
      fromId: this.identity.getUserId(),
      toId: friendId,
      type: 'file-info',
      content,
      timestamp: Date.now(),
      status: 'sent',
    };

    // Send as a regular text message (the metadata is small enough)
    const sessionKey = this.store.getSessionKey(friendId);
    if (!sessionKey) {
      throw new Error(`No session key for friend ${friendId}`);
    }

    const wireData = encryptMessage(
      content,
      sessionKey,
      this.identity.getSigningSecretKey(),
      this.identity.getPublicKey(),
    );

    let delivered = false;
    if (this.p2p.isConnected(friendId)) {
      delivered = this.p2p.send(friendId, Buffer.from(wireData));
    }

    if (!delivered) {
      this.ws.send('message', {
        to: friendId,
        data: {
          type: 'encrypted_message',
          fromId: this.identity.getUserId(),
          payload: Buffer.from(wireData).toString('base64'),
        },
      });
    }

    message.status = delivered ? 'delivered' : 'sent';
    this.store.addMessage(friendId, message);
    this._debouncedSave();

    return message;
  }

  /* ──────────────── Receive ──────────────── */

  /**
   * Decrypt, verify, and store an incoming message.
   *
   * @returns The ChatMessage record, or null if decryption/verification fails.
   */
  receiveMessage(data: Buffer, fromId: string): ChatMessage | null {
    const friendInfo = this.store.friends.get(fromId);
    if (!friendInfo) {
      console.warn(`[ChatManager] Received message from unknown peer: ${fromId}`);
      return null;
    }

    const sessionKey = this.store.getSessionKey(fromId);
    if (!sessionKey) {
      console.warn(`[ChatManager] No session key for peer: ${fromId}`);
      return null;
    }

    const senderPubKey = Buffer.from(friendInfo.publicKey, 'base64');
    const plaintext = decryptMessage(data, sessionKey, senderPubKey);

    if (plaintext === null) {
      console.warn(`[ChatManager] Failed to decrypt message from ${fromId}`);
      return null;
    }

    // Determine message type
    let type: ChatMessage['type'] = 'text';
    try {
      const parsed = JSON.parse(plaintext);
      if (parsed.fileName && parsed.fileSize && parsed.fileHash) {
        type = 'file-info';
      }
    } catch {
      // Not JSON → plain text
    }

    const message: ChatMessage = {
      id: uuidv4(),
      fromId,
      toId: this.identity.getUserId(),
      type,
      content: plaintext,
      timestamp: Date.now(),
      status: 'delivered',
    };

    this.store.addMessage(fromId, message);
    this._debouncedSave();

    return message;
  }

  /* ──────────────── History ──────────────── */

  /** Get chat history for a friend, optionally limited to the last N messages. */
  getChatHistory(friendId: string, limit?: number): ChatMessage[] {
    return this.store.getMessages(friendId, limit);
  }

  /** Delete a specific message from history. */
  deleteMessage(messageId: string): void {
    this.store.deleteMessage(messageId);
    this._debouncedSave();
  }

  /** Mark all messages from a friend as read. */
  markAsRead(friendId: string): void {
    this.store.markAllRead(friendId);
    this._debouncedSave();
  }

  /* ──────────────── Typing indicator ──────────────── */

  /** Send a typing indicator to a friend. */
  sendTyping(friendId: string): void {
    this.ws.send('typing', {
      toId: friendId,
      fromId: this.identity.getUserId(),
    });
  }

  /* ──────────────── Cleanup ──────────────── */

  destroy(): void {
    this.messageCallbacks.length = 0;
    if (this._saveTimer) {
      clearTimeout(this._saveTimer);
      this._saveTimer = null;
      this.store.save();
    }
  }

  /* ──────────────── Debounced save ──────────────── */

  /**
   * Debounced store.save() — coalesces rapid calls into at most one
   * write per 500 ms window, avoiding synchronous I/O on every message.
   */
  private _debouncedSave(): void {
    if (this._saveTimer) return; // already scheduled
    this._saveTimer = setTimeout(() => {
      this._saveTimer = null;
      this.store.save();
    }, 500);
  }
}

File: client/chat_manager.py
"""
Chat message management — encrypt, send, receive, decrypt, store.

Messages are encrypted using the session key established during the
handshake and signed with the identity signing key. P2P transport is
preferred with WebSocket relay fallback.
"""

from __future__ import annotations

import asyncio
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional, Callable

import sys
_shared = str(Path(__file__).resolve().parent.parent / "shared")
if _shared not in sys.path:
    sys.path.insert(0, _shared)

from crypto import (
    encrypt_message,
    decrypt_message,
    encode_base64,
)

from .db import ClientDatabase
from .p2p_client import P2PClient
from .ws_client import WSClient
from .identity_manager import IdentityManager
from .types import ChatMessage, MessageType, MessageStatus, FileMetadata


class ChatManager:
    """Encrypt, send, receive, decrypt, and store chat messages.

    Uses the shared crypto library for encryption/decryption, P2PClient
    for direct peer transport (preferred), and WSClient for relay fallback.
    """

    def __init__(
        self,
        db: ClientDatabase,
        p2p: P2PClient,
        ws: WSClient,
        identity: IdentityManager,
    ):
        self._db = db
        self._p2p = p2p
        self._ws = ws
        self._identity = identity

        self._message_cbs: list[Callable[[ChatMessage], None]] = []
        self._save_timer: Optional[asyncio.Task] = None

        self._setup_listeners()

    # ──────────────── Callbacks ────────────────

    def on_message(self, callback: Callable[[ChatMessage], None]) -> None:
        self._message_cbs.append(callback)

    def _notify_message(self, msg: ChatMessage) -> None:
        for cb in self._message_cbs:
            try:
                cb(msg)
            except Exception:
                pass

    # ──────────────── WebSocket listeners ────────────────

    def _setup_listeners(self) -> None:
        self._ws.on("message", self._on_ws_message)
        self._p2p.on_any_data(self._on_p2p_data)

    async def _on_ws_message(self, msg: dict) -> None:
        data = msg.get("data", {})
        if data.get("type") == "encrypted_message":
            from_id = msg.get("from", data.get("fromId", ""))
            payload = data.get("payload", "")
            import base64
            try:
                raw = base64.b64decode(payload)
            except Exception:
                return
            message = await self.receive_message(raw, from_id)
            if message:
                self._notify_message(message)

    async def _on_p2p_data(self, peer_id: str, data: bytes) -> None:
        message = await self.receive_message(data, peer_id)
        if message:
            self._notify_message(message)

    # ──────────────── Send ────────────────

    async def send_message(self, friend_id: str, content: str) -> ChatMessage:
        """Send a text message to a friend.

        Encrypts with the session key and signs with the identity key.
        Tries P2P first, falls back to WebSocket relay.
        """
        session_key = await self._db.get_session_key(friend_id)
        if not session_key:
            raise RuntimeError(f"No session key for friend {friend_id}")

        now = datetime.now(tz=timezone.utc)
        message = ChatMessage(
            id=str(uuid.uuid4()),
            from_id=await self._identity.get_user_id(),
            to_id=friend_id,
            type=MessageType.TEXT,
            content=content,
            timestamp=now.isoformat(),
            status=MessageStatus.SENT,
        )

        wire_data = encrypt_message(
            content,
            session_key,
            self._identity.get_signing_secret_key(),
            self._identity.get_public_key(),
        )

        delivered = False
        if self._p2p.is_connected(friend_id):
            delivered = self._p2p.send(friend_id, wire_data)

        if not delivered:
            import base64
            self._ws.send("message", {
                "to": friend_id,
                "data": {
                    "type": "encrypted_message",
                    "fromId": await self._identity.get_user_id(),
                    "payload": base64.b64encode(wire_data).decode("ascii"),
                },
            })

        message.status = MessageStatus.DELIVERED if delivered else MessageStatus.SENT
        await self._db.add_message(message)
        return message

    async def send_file_info(self, friend_id: str, file_info: FileMetadata) -> ChatMessage:
        """Send file metadata as a chat message."""
        import json
        content = json.dumps(file_info.to_dict())

        now = datetime.now(tz=timezone.utc)
        message = ChatMessage(
            id=str(uuid.uuid4()),
            from_id=await self._identity.get_user_id(),
            to_id=friend_id,
            type=MessageType.FILE_INFO,
            content=content,
            timestamp=now.isoformat(),
            status=MessageStatus.SENT,
        )

        session_key = await self._db.get_session_key(friend_id)
        if not session_key:
            raise RuntimeError(f"No session key for friend {friend_id}")

        wire_data = encrypt_message(
            content,
            session_key,
            self._identity.get_signing_secret_key(),
            self._identity.get_public_key(),
        )

        delivered = False
        if self._p2p.is_connected(friend_id):
            delivered = self._p2p.send(friend_id, wire_data)

        if not delivered:
            import base64
            self._ws.send("message", {
                "to": friend_id,
                "data": {
                    "type": "encrypted_message",
                    "fromId": await self._identity.get_user_id(),
                    "payload": base64.b64encode(wire_data).decode("ascii"),
                },
            })

        message.status = MessageStatus.DELIVERED if delivered else MessageStatus.SENT
        await self._db.add_message(message)
        return message

    # ──────────────── Receive ────────────────

    async def receive_message(self, data: bytes, from_id: str) -> Optional[ChatMessage]:
        """Decrypt, verify, and store an incoming message."""
        friend = await self._db.get_friend(from_id)
        if not friend:
            print(f"[ChatManager] Received message from unknown peer: {from_id}")
            return None

        session_key = await self._db.get_session_key(from_id)
        if not session_key:
            print(f"[ChatManager] No session key for peer: {from_id}")
            return None

        sender_pub = encode_base64(friend.public_key) if friend.public_key else None
        # For decrypt_message we need the raw public key bytes
        import base64
        try:
            sender_pub_bytes = base64.b64decode(friend.public_key) if friend.public_key else b""
        except Exception:
            sender_pub_bytes = b""

        plaintext = decrypt_message(data, session_key, sender_pub_bytes)
        if plaintext is None:
            print(f"[ChatManager] Failed to decrypt message from {from_id}")
            return None

        # Determine message type
        msg_type = MessageType.TEXT
        try:
            import json
            parsed = json.loads(plaintext)
            if isinstance(parsed, dict) and "fileName" in parsed and "fileHash" in parsed:
                msg_type = MessageType.FILE_INFO
        except Exception:
            pass

        now = datetime.now(tz=timezone.utc)
        message = ChatMessage(
            id=str(uuid.uuid4()),
            from_id=from_id,
            to_id=await self._identity.get_user_id(),
            type=msg_type,
            content=plaintext,
            timestamp=now.isoformat(),
            status=MessageStatus.DELIVERED,
        )

        await self._db.add_message(message)
        return message

    # ──────────────── History ────────────────

    async def get_chat_history(self, friend_id: str, limit: int = 100) -> list[ChatMessage]:
        return await self._db.get_messages(friend_id, limit)

    async def delete_message(self, message_id: str) -> None:
        await self._db.delete_message(message_id)

    async def mark_as_read(self, friend_id: str) -> None:
        await self._db.mark_all_read(friend_id)

    # ──────────────── Typing indicator ────────────────

    def send_typing(self, friend_id: str) -> None:
        self._ws.send("typing", {
            "toId": friend_id,
            "fromId": "",  # Will be filled by WS layer
        })

    # ──────────────── Cleanup ────────────────

    def destroy(self) -> None:
        self._message_cbs.clear()
        if self._save_timer:
            self._save_timer.cancel()
            self._save_timer = None
